python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -1376,7 +1379,7 @@ def should_backfill_atomic_asset_partitions_unit(
                     (
```

Comment:
Re the backfill issue earlier today:

We probably could remove this block as well as `has_partition_mapping_safe_backfill_policy` and instead allow a candidate to run with the parent if:
- They have the same backfill policy
- There is a simple 1:1 partition mapping where the parent partition key is the same as the child's. I think this only applies to the `IdentityPartitionMapping` and the default `TimeWindowPartitionsMapping`

which should fix this multi-asset issue too

Reply:
So, spent a good amount of time on this and agree that the existing state of `has_partition_mapping_safe_backfill_policy` was problematic. However, there is actually a good(ish) reason to add an additional path to being grouped with your parents beyond those first two.

Imagine you have time-partitioned two assets with a fancy lookback partition mapping between them. However, you're actually doing a backfill that targets all partitions of each of those assets. The optimal solution here is to materialize this all in a single run that targets both of these assets (and all partitions for each of them). Even though there's some fancy mapping, we know that all partitions of A will materialize before any B partitions start.

That's what I encapsulated in the new-and-improved additional block of logic. This is much more specific / targeted than before, and should be a lot safer than the previous behavior (which was itself an improvement on previous behavior).

Reply:
Got it -- that makes sense. The new block of logic that handles the "all partitions selected" case looks good to me.